### ISSUE TITLE
Unable to build using Android NDK + CLANG

### ISSUE THREAD

**Issue by sandwichmaker** [flame/blis#207]
I am interested in building BLIS on android and comparing its performance with OpenBLAS and Eigen.

Even though current versions of the android-ndk ship with the GCC compiler as an option, the use of GCC in the NDK is deprecated:

[https://android.googlesource.com/platform/ndk.git/+/master/CHANGELOG.md](https://android.googlesource.com/platform/ndk.git/+/master/CHANGELOG.md)

Currently the BUILD build system when building for ARM explicitly checks for GCC and fails without it.  

But it is straightforward to modify 

config/arm64/make_defs.mk
config/cortexa57/make_defs.mk

to disable the check and modify the compiler flags to work with CLANG. 

At this point we hit a more interesting hurdle

kernels/armv8a/3/bli_gemm_armv8a_opt_4x4.c

which fails with the CLANG complaining 

kernels/armv8a/3/bli_gemm_armv8a_opt_4x4.c:73:1: fatal error: inline assembly requires more registers than available

Poking at the clobber list, it seems the function is using 9 general purpose registers too many. Commenting out any nine general purpose registers makes the function compile (incorrectly of course) and then fail on the next function in the file. 

This is true for NDK version r15c, r16b and r17.

Is this something that can be easily fixed? Also I am curious as to why BLIS is using these rather large functions as inline assembly and not just assembler files?

I am happy to provide more detailed reproduction instructions if that will help.

**Comment by sandwichmaker** [comment]
Passing -fno-integrated-as as recommended by [Clang Migration Notes](https://android.googlesource.com/platform/ndk.git/+/master/docs/ClangMigration.md) does not help either.

**Comment by fgvanzee** [comment]
@sandwichmaker Thanks for this report.

Unfortunately we are light on ARM expertise in the core BLIS development group, so I can't speak for why the code was written to clobber the nine extra registers.

We just added a feature that allows us to modify the default compiler search list based on OS (specifically, the result of `uname -s`) and also fixed a related issue with the compiler being miscommunicated to `make` from `configure`. If you can tell me what `uname -s` returns on ARM, we might be able to isolate your environment and change the search order accordingly, and then you will not have to mess with setting the compiler in `make_defs.mk`.

**Comment by sandwichmaker** [comment]
@fgvanzee this is not a compiler search issue. I am cross compiling for android on linux using the [android-ndk](https://developer.android.com/ndk/downloads/). It is using the right compiler toolchain. 

The settings I had to mess with in make_defs.mk are compiler flags that are GCC specific and are not recognized by CLANG. Thats a minor thing.

The big issue is that the inline assembly compiles with GCC but fails with CLANG and given that GCC support in the NDK is no-more, I would like to compile BLIS with CLANG on ARM. 

Perhaps @figual who seems to be the original author of this kernel (kernels/armv8a/3/bli_gemm_armv8a_opt_4x4.c) can comment?

**Comment by fgvanzee** [comment]
@sandwichmaker I see. Sorry for the misunderstanding. Yes, Francisco (@figual) would be the person to ask. Hopefully he has his github notifications turned on. :) (If he doesn't respond in a day or two, I can send him an email.)

**Comment by figual** [comment]
@sandwichmaker Thanks for your report. I'm looking at it and will come back to you if I need more details, especially with CLANG-related details.

**Comment by fgvanzee** [comment]
@sandwichmaker We've fixed some runtime microkernel bugs in 12dfa95 and 2e31dd7 unrelated to the `clang` vs. `gcc` issue (though they would have needed to get fixed anyway). But @figual may still yet have insights to share regarding the compilation via `clang`.

**Comment by sandwichmaker** [comment]
Thanks for the update @fgvanzee. I appreciate you and @figual looking into it.

**Comment by kali** [comment]
For what it's worth, this is also an issue for iOS, where clang is really the only option.

**Comment by fgvanzee** [comment]
@sandwichmaker I was revisiting this issue to remind myself of the issues at play, and was taking a look at the `dgemm` ukernel for `armv8a`. Can you tell if the problem that you originally reported is due to the fact that clang doesn't want to recognize any general-purpose registers beyond, say, x15?

**Comment by sandwichmaker** [comment]
I tried it again (ndk-19c, blis 0.5.2) , and same error.

```
 blis-0.5.2: make
Generating monolithic blis.h.....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
Generated include/cortexa57/blis.h
Compiling obj/cortexa57/config/cortexa57/bli_cntx_init_cortexa57.o ('cortexa57' CFLAGS for config code)
Compiling obj/cortexa57/kernels/armv8a/3/bli_gemm_armv8a_opt_4x4.o ('cortexa57' CFLAGS for kernels)
kernels/armv8a/3/bli_gemm_armv8a_opt_4x4.c:78:1: fatal error: inline assembly requires more registers than available
"                                            \n\t"
^
1 error generated.
make: *** [obj/cortexa57/kernels/armv8a/3/bli_gemm_armv8a_opt_4x4.o] Error 1
```

**Comment by fgvanzee** [comment]
@sandwichmaker Okay, thanks. I will talk to our resident ARM expert, @dnparikh, about whether we might be able to rewrite the kernel to use fewer registers. If that is the only error (rather than the first of many), then we might be able to solve this without going to intrinsics or straight assembly code.

**Comment by sandwichmaker** [comment]
Thanks @fgvanzee . Please let me know if I can do anything to help test/replicate.

**Comment by devinamatthews** [comment]
This should be fixed by #504 and #492.
